### PR TITLE
Scheduler: export node field in NodeInfo

### PR DESCRIPTION
**Pull request by gavinfish**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

pre-PR for #94696, export node field in NodeInfo so that it can be used with cmp directly.

**Which issue(s) this PR fixes**:

Xref #94696

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by ahg-g** [comment]
I suggest you split the change to kubelet into a separate PR so that the change we send to their approvals is as small as possible.

File: pkg/scheduler/framework/types.go [hunk 1]
```diff
@@ -467,18 +467,10 @@ func NewNodeInfo(pods ...*v1.Pod) *NodeInfo {
 	return ni
 }
 
-// Node returns overall information about this node.
-func (n *NodeInfo) Node() *v1.Node {
-	if n == nil {
```
Comment (adtac) [inline pkg/scheduler/framework/types.go:472]: you probably already thought of this, but do we know if we call `Node()` on a nil struct anywhere? if we do, such a call would safely return nil before, but with this change we'd get a nil deref panic

Comment (gavinfish) [reply pkg/scheduler/framework/types.go:472]: yeah, this is the most effort for this PR, having to check all the callers and find out where do they get the nodeInfo struct. Most of them are from snapshot and few from construction. So I think it should fine.

**Comment by gavinfish** [comment]
> I suggest you split the change to kubelet into a separate PR so that the change we send to their approvals is as small as possible.

Agree that we'd better reduce cross-sig PRs. The problem here is that we cannot keep `Node` field and `Node()` func at the same time. So I cannot create two individual PRs for scheduling and kubelet. 

To reduce kubelet sig reviewing work, another option is:
1. A PR creating a temporary func `GetNode()` to replace all `Node()` in scheduling
2. A PR to export `Node` and remove `Node()` func for kubelet
3. A PR to cleanup `GetNode()` in scheduling

**Comment by ahg-g** [comment]
> > I suggest you split the change to kubelet into a separate PR so that the change we send to their approvals is as small as possible.
> 
> Agree that we'd better reduce cross-sig PRs. The problem here is that we cannot keep `Node` field and `Node()` func at the same time. So I cannot create two individual PRs for scheduling and kubelet.
> 
> To reduce kubelet sig reviewing work, another option is:
> 
> 1. A PR creating a temporary func `GetNode()` to replace all `Node()` in scheduling
> 2. A PR to export `Node` and remove `Node()` func for kubelet
> 3. A PR to cleanup `GetNode()` in scheduling

that is too much, just try to have the node related changes in a separate commit.

**Comment by gavinfish** [comment]
> > > I suggest you split the change to kubelet into a separate PR so that the change we send to their approvals is as small as possible.
> > 
> > 
> > Agree that we'd better reduce cross-sig PRs. The problem here is that we cannot keep `Node` field and `Node()` func at the same time. So I cannot create two individual PRs for scheduling and kubelet.
> > To reduce kubelet sig reviewing work, another option is:
> > 
> > 1. A PR creating a temporary func `GetNode()` to replace all `Node()` in scheduling
> > 2. A PR to export `Node` and remove `Node()` func for kubelet
> > 3. A PR to cleanup `GetNode()` in scheduling
> 
> that is too much, just try to have the node related changes in a separate commit.

Done. Please take a look for scheduling part.
/assign @ahg-g

File: pkg/scheduler/internal/cache/cache.go [hunk 2]
```diff
@@ -226,7 +226,7 @@ func (cache *schedulerCache) UpdateSnapshot(nodeSnapshot *Snapshot) error {
 			// Transient scheduler info is reset here.
 			node.info.TransientInfo.ResetTransientSchedulerInfo()
 		}
-		if np := node.info.Node(); np != nil {
+		if np := node.info.Node; np != nil {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/cache/cache.go:229]: this is the most critical change.
If we are sure that `node.info != nil` at this point, then we can safely change usages in plugins and what not. Can you double check and give a summary of where are these objects coming from?

Comment (gavinfish) [reply pkg/scheduler/internal/cache/cache.go:229]: When nodes/pods added or updated, we will use `n = newNodeInfoListItem(framework.NewNodeInfo())` to init cache node which assigns the new NodeInfo to `node.info`.

Another trick way to support `node.info != nil` at this point is that if so, line 221 will raise error.

Comment (alculquicondor) [reply pkg/scheduler/internal/cache/cache.go:229]: /approve
/lgtm

thanks for checking

**Comment by gavinfish** [comment]
/assign @derekwaynecarr
for kubelet

**Pull request closed without merging**